xmppserver/src/main/java/org/jivesoftware/util/SAXReaderUtil.java

```diff
@@ -0,0 +1,287 @@
+
+    public static final SystemProperty<Integer> PARSER_SERVICE_MAX_POOL_SIZE = SystemProperty.Builder.ofType(Integer.class)
+                                                                                    .setKey("xmpp.xmlutil.parser.maximum-pool-size")
+                                                                                    .setMinValue(1)
+                                                                                    .setDefaultValue(25)
+                                                                                    .setDynamic(false)
+                                                                                    .build();
+
+    public static final SystemProperty<Duration> PARSER_SERVICE_KEEP_ALIVE_TIME = SystemProperty.Builder.ofType(Duration.class)
+                                                                                    .setKey("xmpp.xmlutil.parser.keep_alive_time")
+                                                                                    .setChronoUnit(ChronoUnit.SECONDS)
+                                                                                    .setMinValue(Duration.ofMillis(0))
+                                                                                    .setDefaultValue(Duration.ofMinutes(1))
+                                                                                    .setDynamic(false)
+                                                                                    .build();
+
+    private static final ThreadPoolExecutor parserService = new ThreadPoolExecutor(
+                                                            PARSER_SERVICE_CORE_POOL_SIZE.getValue(),
+                                                            PARSER_SERVICE_MAX_POOL_SIZE.getValue(),
+                                                            PARSER_SERVICE_KEEP_ALIVE_TIME.getValue().toMillis(), TimeUnit.MILLISECONDS,
+                                                            new SynchronousQueue<>());
+
+    static
+    {
+        if (JMXManager.isEnabled()) {
+            final ThreadPoolExecutorDelegateMBean mBean = new ThreadPoolExecutorDelegate(parserService);
+            JMXManager.tryRegister(mBean, ThreadPoolExecutorDelegateMBean.BASE_OBJECT_NAME + "saxReaderUtil");
+        }
+    }
+
+    /**
+     * Parses an XML Document from the content provided by an input stream.
+     *
+     * The XML parsing is delegated to a limited set of parsers. When a parser is not immediately available, this method
+     * will block until one becomes available. The availability of parsers can be optimized by using the SystemProperties
+     * in this class.
+     *
+     * @param stream The data to be parsed as an XML Document.
+     * @return an XML Document.
+     */
+    public static Document readDocument(@Nonnull final InputStream stream) throws ExecutionException, InterruptedException {
+        return parserService.submit(new ParserTask(stream)).get();
```

Comment:
I'm surprised that an IOException isn't being thrown here, as you're reading from an InputStream - have the underlying exceptions been lost by the use of a pool?

Reply:
The 'read' action wraps any exceptions (including an IOException) in a proprietary DocumentException, but your argument stands.

By feeding it through an executor service _as a Callable_, any Exceptions are wrapped in an ExecutionException - so they're not lost, but transformed. I've pushed a test in a2f806475eb8f5b8123335ec6b639e948e9e755f that verifies this behavior.

The API that I introduced is blocking - maybe it makes sense to 'unwrap' this exception (checking if the cause of the ExecutionException is a DocumentException, and if so, rethrow that), to have a more consistent signature to the method, as compared to the 'old' scenario. I'm not sure though if such juggling would end up causing more confusion than fixing any. What do you think @GregDThomas?

Reply:
The problem is you have synchronous functionality ("parse this") wrapped in a thread pool so throwing asynchronous related exceptions. The callers don't really care about about the async problems, they just want the answer or an exception thrown, and may deal with different exceptions differently (badly formed packets vs. broken stream, for example)

As such, I think the `ExecutionException` loses the meaning of any errors. Perhaps unwrap `SAXException`, `DocumentException` from that and make those explicit throws. Looking at the code, I /think/ anything else must be a `RuntimeException`, so you could unwrap those too, and perhaps re-throw anything else wrapped in a `RuntimeException` (I don't think there can be any other exceptions bar the ones listed, but 🤷).

I do also wonder if you could catch the `InterruptedException`, re-interrupt the thread and re-throw as a some sort of `RuntimeException` (perhaps `IllegalStateException`) to save every caller having to deal with it?

Reply:
I see your point - it's pretty much what I was trying to get at in my earlier comment, but having given it some more thought, I don't think we should go down that road. By unwrapping the cause of the ExecutionException we'd be exchanging one awkward solution with one that might be equally awkward, but in different places. We'd end up creating an implementation that's:

- a bit more complex / harder to maintain due to the exception juggling;
- strange to have to debug ("why did I lose part of my stacktrace?");
- still not a 100% fit (particularly with regards to there being a need to throw unexpected Exceptions in a new RuntimeException, which feels somewhat of a concession), leaving room for additional awkwardness.

The article that you referenced earlier (with regards to setting the interrupted flag) explicitly mentioned to _not_ rethrow an InterruptedException as a RuntimeException, for very similar reasoning as what you're using here: details would be hidden. In the case of an interrupt, even if we'd interrupt a async facility that we are trying to hide, that's undesirable.

The benefit of unwrapping is that the API would more explicitly define the type of exception - but even without this unwrapping, those types would still be visible (albeit as a cause of the exceptions thrown, instead of as the exceptions themselves). So, even though not very pretty, we're not losing information.

Although I don't quite like exposing some of the artifacts of having this being fed through an ExecutorService, trying to hide that won't be much better, especially since we'd have to make concessions there too. Maybe the ugliness of the solution is more palatable by adding async methods (that return a future, instead of blocking until done)?

Reply:
I think I agree with everything you've written. Although it's an API (by nature of the `public` nature of the methods) it's not really a library you're writing, so it's probably less important that you hide the complexities of what's going on behind the scenes. So happy to close this one off :)

Reply:
I've now added an async facility. The original blocking methods now, in javadoc, refer to the fact that they're wrappers for the async equivalents. At least in my head, that makes a difference. :)